### ISSUE TITLE
mark disappears, if a transformation has the same value as used by default

### ISSUE THREAD

**Issue by nyurik** [vega/vega#1016]
```
{
  "$schema": "https://vega.github.io/schema/vega/v3.0.json",
  "data": [
    {
      "name": "table",
      "values": [
        {
          "foo": 6,
          "y": 50,
          "text": "something"
        }
      ]
    }
  ],
  "marks": [
    {
      "type": "text",
      "from": {
        "data": "table"
      },
      "transform": [
        {

*************  rename from "x" to "xx", and it will work ****************
          "type": "formula", "expr": "datum.foo + 10", "as": "x"      
        }
      ],
      "encode": {
        "update": {
          "x": {"value": 100},
          "y": {"value": 100},
          "text": { "value": "raw text" }
        }
      }
    }
  ]
}```

**Comment by nyurik** [comment]
There might also be some inconsistencies when mark skips the x/y parameters, but instead relies on the defaults -- they work if set by the transform inside the mark, but not if set by transform at the data, or contained in the data itself.

**Comment by jheer** [comment]
Here is what's happening in this example:

1. The dataflow runs for the first time. The `update` is executed before the `formula`, as one would expect. The resulting item should have an x-coordinate of 16 (= 6 + 10).
2. Auto-sizing is performed (here, `"pad"`) based on the computed values. This determines the view width and height. As the width/height signals default to `-1` when not specified, they end up getting updated to `0` (with the rest of the sizing treated as padding).
3. Because the width/height signals have been updated, Vega schedules an additional execution of the dataflow graph. However, during this execution the `formula` transform exits early as none of the dependent fields (here `foo`) have changed. This results in the scenegraph item being set back to `100` (as this value is never over-written).
4. As a result of the above, the view size has been determined based on the `formula` x value, but the text item is later repositioned based on the `update` x value. Hence it "disappears".

I am exploring two fixes here:
1. Have undefined width/height signals default to 0, not -1. This makes more sense and prevents an unnecessary re-execution of the dataflow graph. In this case, this change is sufficient to make the error go away (at least initially). It does not, however, fix the more general issue of different calculations overwriting each other.
2. Have the formula transform re-run (i.e., don't exit early) if the output field has been modified (not just the dependent input fields). However, I'm a little concerned about this step more generally. While it may make sense for the formula transform in this case, does it make sense in all other cases? Also, should *all* transforms similarly include their output fields for checks? I'm not sure this would work out well.

As a result, I plan to go ahead with the first fix now, but will need to do more analysis (at a future, unspecified date) to see how safe/correct the second fix is.

For the most part, it is considered "bad form" to have formula transforms overwrite values of the original data or of those output by other transforms or encoders. While technically allowed by the schema and parser, it is also not advisable to have formula transforms generate scenegraph properties for mark items. For visual encoding it is instead preferred to use a signal expression as part of an encode block.

**Comment by jheer** [comment]
The fix for width/height signals has been released in Vega 3.0.3. If having formula (or other) transforms re-run in response to changed _output_ fields is something that is sufficiently important, please open a new issue along with an argument for why the feature should be added (for example, as opposed to using signal expressions in encoding blocks or other potential designs). Thanks!
